dex/src/main/java/io/crate/data/AsyncCompositeBatchIterator.java

```diff
@@ -101,29 +108,41 @@ public void close() {
         if (allLoaded()) {
             return CompletableFutures.failedFuture(new IllegalStateException("BatchIterator already loaded"));
         }
-
-        List<CompletableFuture<CompletableFuture<?>>> asyncInvocationFutures = new ArrayList<>();
-        for (BatchIterator iterator : iterators) {
-            if (iterator.allLoaded()) {
-                continue;
+        int availableThreads = this.availableThreads.getAsInt();
+        List<BatchIterator<T>> itToLoad = getUnloadedIterators();
+
+        List<CompletableFuture<CompletableFuture>> nestedFutures = new ArrayList<>();
+        if (availableThreads < itToLoad.size()) {
+            Iterable<List<BatchIterator<T>>> iteratorsPerThread = Iterables.partition(
+                itToLoad, itToLoad.size() / availableThreads);
+
+            for (List<BatchIterator<T>> batchIterators: iteratorsPerThread) {
+                CompletableFuture<CompletableFuture> future = supplyAsync(() -> {
+                    ArrayList<CompletableFuture<?>> futures = new ArrayList<>(batchIterators.size());
+                    for (BatchIterator<T> batchIterator: batchIterators) {
+                        futures.add(batchIterator.loadNextBatch().toCompletableFuture());
+                    }
+                    return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+                }, executor);
+                nestedFutures.add(future);
+            }
+        }  else {
+            for (BatchIterator<T> iterator: itToLoad) {
+                nestedFutures.add(supplyAsync(() -> iterator.loadNextBatch().toCompletableFuture(), executor));
             }
-            Supplier<CompletableFuture<?>> loadNextBatchFuture = () -> iterator.loadNextBatch().toCompletableFuture();
-
-            // Every "supplyAsync future" completes with the iterator::loadNextBatch() future, so in order to
-            // access the loadNextBatch future, the asyncInvocation one should be completed and the result extracted
-            CompletableFuture<CompletableFuture<?>> supplyAsyncFuture =
-                CompletableFuture.supplyAsync(loadNextBatchFuture, executor);
-            asyncInvocationFutures.add(supplyAsyncFuture);
         }
+        return CompletableFutures.allAsList(nestedFutures)
+            .thenCompose(innerFutures -> CompletableFuture.allOf(innerFutures.toArray(new CompletableFuture[0])));
+    }
 
-        CompletableFuture<List<CompletableFuture<?>>> loadNextBatchListFuture =
-            CompletableFutures.allAsList(asyncInvocationFutures);
-
-        return loadNextBatchListFuture.thenCompose(
-            (List<CompletableFuture<?>> loadNextBatchFutures) ->
-                // Future that'll wait for all loadNextBatch futures to complete
-                CompletableFuture.allOf(loadNextBatchFutures.toArray(new CompletableFuture[0]))
-        );
+    private List<BatchIterator<T>> getUnloadedIterators() {
```

Comment:
this could receive the iterators as an argument and become a static method

Maybe it's just me but "unloaded" in the iterator context is a bit confusing. I'd rather have a more verbose method name ( `getIteratorsWithUnloadedBatches` / `iteratorsWithRemainingBatchesToLoad` or something along those lines - does "unexhausted" make sense to you? maybe that could be a concept)

Reply:
> this could receive the iterators as an argument and become a static method

I know it could, but why would that be better?

How about `getIteratorsToLoad`

Reply:
+1 on `getIteratorsToLoad`

I think having the method act as a function is easier to reason about as the scope is reduced to the method definition